### ISSUE TITLE
API for multi-criteria filtering

### ISSUE THREAD

**Issue by rvansa** [lukehutch/fast-classpath-scanner#69]
Hi, is it possible to do multi-criteria filtering (class implements interface X and is annotated by Y)? I would think about stream of ClassInfos (or similar descriptors).

**Comment by lukehutch** [comment]
Hi @rvansa, this is a really interesting idea. I currently have some API calls like .matchInterfacesAllOf(...), but it would be better to have something like .matchClasses().thatImplement(...).and().thatExtend(...). I can port all the existing API calls to use this matching form internally.

I'm ramping up to do a 2.0 release in the near future, so I'll put this on my TODO list. Thanks for the suggestion!

**Comment by lukehutch** [comment]
PS @rvansa, if you have (or if anyone else has) suggestions about how the API should look, please leave them here. Thanks!

**Comment by lukehutch** [comment]
And actually, the ClassInfo objects are not returned to the user currently, but as you suggest, maybe it would be an interesting thing to return these so that users can see all the info about a returned class without having to resort to introspection.

**Comment by rvansa** [comment]
Fluent DSL API sounds like overkill to me, but I've just started to use FCS. Streams are a standard way of prefiltering any results, so I would be content with

```java
new FCS().stream()
    .filter(ci -> ci.isAnnotationPresent(annotationName))
    .filter(ci -> ci.isAssignableTo(superClassName))
    .map(ClassInfo::toClass).collect(Collectors.toList())
```

**Comment by lukehutch** [comment]
Ah, you mean Java 8 Streams. Unfortunately a number of users of FCS need it to work on Java 7. Maybe I should start a new library that adds Java 8 support to FCS?

**Comment by lukehutch** [comment]
Hi @rvansa, I implemented this after realizing I don't need to support Java 8 for the API to work with Java 8 streams -- see the commits above.

Please let me know if this does what you want. It was a great suggestion!

**Comment by lukehutch** [comment]
PS you need to call .values() before .stream(), see:

https://github.com/lukehutch/fast-classpath-scanner/blob/master/src/test/java/io/github/lukehutch/fastclasspathscanner/test/Java8Test.java

**Comment by rvansa** [comment]
Cool, that was really fast fix! Thanks - I'll use this as soon as there is a release in Maven.
(btw. yesterday I've closed the issue by mistake, but now it probably can be closed).

**Comment by lukehutch** [comment]
You're welcome. Documented and released in 1.93.0. Please test and let me know if it works for you.
